src/slice.rs

```diff
@@ -282,188 +303,323 @@ impl_sliceorindex_from_index!(isize);
+
+    fn out_ndim(&self) -> usize;
+}
+
+macro_rules! impl_canslice_samedim {
+    ($in_dim:ty) => {
+        unsafe impl<T, Do> CanSlice<$in_dim> for SliceInfo<T, $in_dim, Do>
+        where
+            T: AsRef<[AxisSliceInfo]>,
+            Do: Dimension,
+        {
+            type OutDim = Do;
+
+            fn in_ndim(&self) -> usize {
+                self.in_ndim()
+            }
+
+            fn out_ndim(&self) -> usize {
+                self.out_ndim()
+            }
+        }
+    };
+}
+impl_canslice_samedim!(Ix0);
+impl_canslice_samedim!(Ix1);
+impl_canslice_samedim!(Ix2);
+impl_canslice_samedim!(Ix3);
+impl_canslice_samedim!(Ix4);
+impl_canslice_samedim!(Ix5);
+impl_canslice_samedim!(Ix6);
+
+unsafe impl<T, Di, Do> CanSlice<IxDyn> for SliceInfo<T, Di, Do>
+where
+    T: AsRef<[AxisSliceInfo]>,
+    Di: Dimension,
+    Do: Dimension,
+{
+    type OutDim = Do;
+
+    fn in_ndim(&self) -> usize {
+        self.in_ndim()
+    }
+
+    fn out_ndim(&self) -> usize {
+        self.out_ndim()
+    }
+}
+
 /// Represents all of the necessary information to perform a slice.
 ///
-/// The type `T` is typically `[SliceOrIndex; n]`, `[SliceOrIndex]`, or
-/// `Vec<SliceOrIndex>`. The type `D` is the output dimension after calling
-/// [`.slice()`].
+/// The type `T` is typically `[AxisSliceInfo; n]`, `[AxisSliceInfo]`, or
+/// `Vec<AxisSliceInfo>`. The type `Di` is the dimension of the array to be
+/// sliced, and `Do` is the output dimension after calling [`.slice()`]. Note
+/// that if `Di` is a fixed dimension type (`Ix0`, `Ix1`, `Ix2`, etc.), the
+/// `SliceInfo` instance can still be used to slice an array with dimension
+/// `IxDyn` as long as the number of axes matches.
 ///
 /// [`.slice()`]: struct.ArrayBase.html#method.slice
 #[derive(Debug)]
 #[repr(C)]
-pub struct SliceInfo<T: ?Sized, D: Dimension> {
-    out_dim: PhantomData<D>,
+pub struct SliceInfo<T: ?Sized, Di: Dimension, Do: Dimension> {
+    in_dim: PhantomData<Di>,
+    out_dim: PhantomData<Do>,
     indices: T,
 }
 
-impl<T: ?Sized, D> Deref for SliceInfo<T, D>
+impl<T: ?Sized, Di, Do> Deref for SliceInfo<T, Di, Do>
 where
-    D: Dimension,
+    Di: Dimension,
+    Do: Dimension,
 {
     type Target = T;
     fn deref(&self) -> &Self::Target {
         &self.indices
     }
 }
 
-impl<T, D> SliceInfo<T, D>
+impl<T, Di, Do> SliceInfo<T, Di, Do>
 where
-    D: Dimension,
+    Di: Dimension,
+    Do: Dimension,
 {
     /// Returns a new `SliceInfo` instance.
     ///
-    /// If you call this method, you are guaranteeing that `out_dim` is
-    /// consistent with `indices`.
+    /// If you call this method, you are guaranteeing that `in_dim` and
+    /// `out_dim` are consistent with `indices`.
     #[doc(hidden)]
-    pub unsafe fn new_unchecked(indices: T, out_dim: PhantomData<D>) -> SliceInfo<T, D> {
+    pub unsafe fn new_unchecked(
+        indices: T,
+        in_dim: PhantomData<Di>,
+        out_dim: PhantomData<Do>,
+    ) -> SliceInfo<T, Di, Do> {
```

Comment:
This function indent, paranthesis and brace style, what is this, is this what rustfmt prefers?

Reply:
Yes, it's the `rustfmt` style when the function/method params don't fit on one line. (See the [style guide](https://github.com/rust-lang/rfcs/blob/master/style-guide/items.md#function-definitions).)